### PR TITLE
warn wallet_watchAsset ERC721,ERC1155

### PR DESCRIPTION
**Pull request by jiexi**
Adds a console.warn when `wallet_watchAsset` is called with `params: {type: 'ERC721'}` or `params: {type: 'ERC1155'}`.

### CODE REVIEW

File: src/middleware/createRpcWarningMiddleware.ts [hunk 1]
```diff
@@ -27,6 +28,15 @@ export function createRpcWarningMiddleware(
     ) {
       log.warn(messages.warnings.rpc.ethGetEncryptionPublicKeyDeprecation);
       sentWarnings.ethGetEncryptionPublicKeyDeprecation = true;
+    } else if (
+      !sentWarnings.walletWatchAssetNFTExperimental &&
+      req.method === 'wallet_watchAsset' &&
+      ['ERC721', 'ERC1155'].includes(
```
Comment (jiexi) [inline src/middleware/createRpcWarningMiddleware.ts]: should these get pulled in as constants from controller-utils?

File: src/messages.ts [hunk 2]
```diff
@@ -37,6 +37,7 @@ const messages = {
     rpc: {
       ethDecryptDeprecation: `MetaMask: The RPC method 'eth_decrypt' is deprecated and may be removed in the future.\nFor more information, see: https://medium.com/metamask/metamask-api-method-deprecation-2b0564a84686`,
       ethGetEncryptionPublicKeyDeprecation: `MetaMask: The RPC method 'eth_getEncryptionPublicKey' is deprecated and may be removed in the future.\nFor more information, see: https://medium.com/metamask/metamask-api-method-deprecation-2b0564a84686`,
+      walletWatchAssetNFTExperimental: `MetaMask: The RPC method 'wallet_watchAsset' is experimental for ERC721/ERC1155 assets and may change in the future.\nFor more information, see: https://TODO`,
```
Comment (jiexi) [inline src/messages.ts]: what should this messaging be?

Comment (shanejonas) [reply src/messages.ts]: `see: https://docs.metamask.io/wallet/reference/wallet_watchasset/` maybe?

Comment (adonesky1) [reply src/messages.ts]: maybe we link to [MIP-1](https://github.com/MetaMask/metamask-improvement-proposals/blob/main/MIPs/mip-1.md) and the [proposal lifecycle](https://github.com/MetaMask/metamask-improvement-proposals/blob/main/PROCESS-GUIDE.md#proposal-lifecycle) in the process-guide document? cc @vandan

File: src/middleware/createRpcWarningMiddleware.ts [hunk 3]
```diff
@@ -27,6 +28,15 @@ export function createRpcWarningMiddleware(
...
+        (req as JsonRpcRequest<{ type: string }>).params?.type || '',
```
Comment (adonesky1) [inline src/middleware/createRpcWarningMiddleware.ts:36]: do we need `|| ''` here?

**Review by adonesky1 — COMMENTED** [review]
One small question, and we just need to update that warning message and then I think we're good to go!

File: src/middleware/createRpcWarningMiddleware.ts (hunk 3 above)
Comment (jiexi) [reply src/middleware/createRpcWarningMiddleware.ts:36]: `Array<string>.includes` expects only `string` type as an arg. it's potentially undefined due to the `?` optional chaining

Comment (adonesky1) [reply src/middleware/createRpcWarningMiddleware.ts:36]: ah ok makes sense

File: src/messages.ts (hunk 2 above)
Comment (vandan) [reply src/messages.ts]: Yes, I think both of those links would be good references.

**Comment by jiexi** [comment]
[@metamask/controller-utils@4.0.0](https://socket.dev/npm/package/@metamask/controller-utils/overview/4.0.0) is used in extension. Okay if we `@SocketSecurity ignore-all` the socket-security alerts?

File: src/middleware/createRpcWarningMiddleware.test.ts [hunk 4]
```diff
@@ -79,51 +112,56 @@ describe('createRpcWarningMiddleware', () => {
...
-    it('should not issue a warning', async () => {
-      const consoleWarnSpy = jest.spyOn(globalThis.console, 'warn');
-      const middleware = createRpcWarningMiddleware(globalThis.console);
-      const engine = new JsonRpcEngine();
-      engine.push(middleware);
-
-      await engine.handle({ jsonrpc: '2.0', id: 1, method: 'eth_chainId' });
+  for (const { method, params = {}, scenario } of unaffected) {
+    const name = scenario ? `${method} (${scenario})` : method;
+    describe(`${name}`, () => {
```
Comment (adonesky1) [inline src/middleware/createRpcWarningMiddleware.test.ts]: [nit] looks like you could use [describe.each](https://jestjs.io/docs/api#1-describeeachtablename-fn-timeout) here.

File: package.json [hunk 5]
```diff
@@ -37,6 +37,7 @@
     "test:watch": "jest --watch"
   },
   "dependencies": {
+    "@metamask/controller-utils": "^4.0.0",
```
Comment (adonesky1) [inline package.json]: All of the added sub-deps here will be deduped in the extension but I'm wondering its worth making this package so much heavier just for those asset type strings... Thoughts @mcmire @Gudahtt ?

Comment (adonesky1) [reply package.json]: I realize I originally said yes when you asked if we should pull the consts in this way...

Comment (jiexi) [reply package.json]: same thoughts about if it is worth including the package

**Review by pedronfigueiredo — DISMISSED** [review]

File: package.json (hunk 5 above)
Comment (mcmire) [reply package.json]: Hmm, I also have the same qualms. We should definitely try to make sure that our libraries are as small as possible on their own. Perhaps we just copy over those constants for now.

**Review by adonesky1 — APPROVED** [review]
LGTM!

**Pull request merged**